src/fides/api/util/tcf/tc_string.py

```diff
@@ -180,3 +188,87 @@ def build_disclosed_vendors_string(model: TCModel) -> str:
+        preference: bool = tc_string_preferences.get(identifier, False)
+
+        if isinstance(preference_class, TCFVendorSave):
+            # Vendors are currently saved as strings in our db
+            identifier = str(identifier)
+
+        fides_preference: FidesPreferenceType = preference_class(
+            id=identifier, preference=boolean_to_user_consent_preference(preference)
+        )
+        preferences_array.append(fides_preference)
+    return preferences_array
+
+
+def decode_tc_string_to_preferences(
+    tc_string: str, tcf_contents: TCFExperienceContents
+) -> TCStringFidesPreferences:
+    """Method to convert a TC String into a TCStringFidesPreferences object, which is a format from which
+    preferences can be saved into the Fides database"""
+    try:
+        # Decode the string and pull the user opt-ins off of the string
+        decoded: ConsentV2 = decode_v2(tc_string)
+        tc_str_p_c: Dict[int, bool] = decoded.purposes_consent
+        tc_str_p_li: Dict[int, bool] = decoded.purposes_legitimate_interests
+        tc_str_v_c: Dict[int, bool] = decoded.consented_vendors
+        tc_str_v_li: Dict[int, bool] = decoded.interests_vendors
+        tc_str_sf: Dict[int, bool] = decoded.special_features_optin
+    except (binascii.Error, AttributeError):
+        raise DecodeTCStringError("Invalid base64-encoded TC string")
+
+    # From our datamap, build all the possible options for the TC string, if the user
+    # opted into everything
+    all_options_tc_model: TCModel = convert_tcf_contents_to_tc_model(
+        tcf_contents, UserConsentPreference.opt_in
+    )
+    datamap_p_c: List[int] = all_options_tc_model.purpose_consents
+    datamap_p_li: List[int] = all_options_tc_model.purpose_legitimate_interests
+    datamap_v_c: List[int] = all_options_tc_model.vendor_consents
+    datamap_v_li: List[int] = all_options_tc_model.vendor_legitimate_interests
+    datamap_sf: List[int] = all_options_tc_model.special_feature_optins
+
+    # Return TCString Preferences that are driven by the datamap.  For every element in the datamap,
+    # consider it an opt-in preference if it's included in the TC string, otherwise, consider
+    # it an opt-out preference.
+    return TCStringFidesPreferences(
+        purpose_consent_preferences=convert_to_fides_preference(
+            datamap_p_c, tc_str_p_c, TCFPurposeSave
+        ),
+        purpose_legitimate_interests_preferences=convert_to_fides_preference(
+            datamap_p_li, tc_str_p_li, TCFPurposeSave
+        ),
+        vendor_consent_preferences=convert_to_fides_preference(
+            datamap_v_c, tc_str_v_c, TCFVendorSave
+        ),
+        vendor_legitimate_interests_preferences=convert_to_fides_preference(
+            datamap_v_li, tc_str_v_li, TCFVendorSave
+        ),
+        special_feature_preferences=convert_to_fides_preference(
+            datamap_sf, tc_str_sf, TCFSpecialFeatureSave
+        ),
+    )
```

Comment:
Note that the TC string only encodes a subset of the sections that we supply in the TCF overlay, so I can only save preferences for purpose consent, purpose legitimate interests, vendor consent, vendor legitimate interests, and special features.